// steel-sdk/resources/sessions.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionListResponsesSessionsCursor = exports.Sessions = void 0;
const resource_1 = require("../resource.js");
const core_1 = require("../core.js");
const pagination_1 = require("../pagination.js");
class Sessions extends resource_1.APIResource {
    create(body = {}, options) {
        if ((0, core_1.isRequestOptions)(body)) {
            return this.create({}, body);
        }
        return this._client.post('/v1/sessions', { body, ...options });
    }
    /**
     * Retrieves details of a specific session by ID.
     */
    retrieve(id, options) {
        return this._client.get(`/v1/sessions/${id}`, options);
    }
    list(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/v1/sessions', SessionListResponsesSessionsCursor, { query, ...options });
    }
    /**
     * Fetches the context data of a specific session.
     */
    context(id, options) {
        return this._client.get(`/v1/sessions/${id}/context`, options);
    }
    /**
     * Releases a specific session by ID.
     */
    release(id, options) {
        return this._client.post(`/v1/sessions/${id}/release`, options);
    }
    /**
     * Releases all active sessions for the current organization.
     */
    releaseAll(options) {
        return this._client.post('/v1/sessions/release', options);
    }
}
exports.Sessions = Sessions;
class SessionListResponsesSessionsCursor extends pagination_1.SessionsCursor {
}
exports.SessionListResponsesSessionsCursor = SessionListResponsesSessionsCursor;
Sessions.SessionListResponsesSessionsCursor = SessionListResponsesSessionsCursor;
